nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -1583,6 +1626,274 @@ impl DataStore {
+                        })?;
+
+                    // Set the replacement region's volume id to the target's
+                    // volume id
+                    diesel::update(region_dsl::region)
+                        .filter(region_dsl::id.eq(replacement.region_id))
+                        .set(region_dsl::volume_id.eq(target.volume_id))
+                        .execute_async(&conn)
+                        .await
+                        .map_err(|e| {
+                            err.bail_retryable_or_else(e, |e| {
+                                VolumeReplaceRegionError::Public(
+                                    public_error_from_diesel(
+                                        e,
+                                        ErrorHandler::Server,
+                                    )
+                                )
+                            })
+                        })?;
+
+                    // Update the target volume's construction request to
+                    // replace the target region's SocketAddrV6 with the
+                    // replacement region's
+                    let maybe_old_volume = {
+                        volume_dsl::volume
+                            .filter(volume_dsl::id.eq(target.volume_id))
+                            .select(Volume::as_select())
+                            .first_async::<Volume>(&conn)
+                            .await
+                            .optional()
+                            .map_err(|e| {
+                                err.bail_retryable_or_else(e, |e| {
+                                    VolumeReplaceRegionError::Public(
+                                        public_error_from_diesel(
+                                            e,
+                                            ErrorHandler::Server,
+                                        )
+                                    )
+                                })
+                            })?
+                    };
+
+                    let old_volume = if let Some(old_volume) = maybe_old_volume {
+                        old_volume
+                    } else {
+                        // Target volume was deleted, so return an error, we
+                        // can't perform the region replacement now!
+                        return Err(err.bail(VolumeReplaceRegionError::TargetVolumeDeleted));
+                    };
+
+                    let old_vcr: VolumeConstructionRequest =
+                        match serde_json::from_str(&old_volume.data()) {
+                            Ok(vcr) => vcr,
+                            Err(e) => {
+                                return Err(err.bail(VolumeReplaceRegionError::SerdeError(e)));
+                            },
+                        };
+
+                    // Copy the old volume's VCR, changing out the old region
+                    // for the new.
+                    let new_vcr = match replace_region_in_vcr(
```

Comment:
If this transaction retries, the generation will keep getting bumped? Is that intended?

Reply:
It is yeah. If it retries, one of the conflicts that can cause that is concurrent modification of the volume, and whatever order volume modifications get applied in the generation should be bumped so that the last written has the highest version.